s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/input/AvroTransformer.java

```diff
@@ -66,18 +66,67 @@ public byte[] getValueBytes(final Object record, final String topic, final S3Sou
 
     private Stream<Object> readAvroRecordsAsStream(final IOSupplier<InputStream> inputStreamIOSupplier,
             final DatumReader<GenericRecord> datumReader) {
-        try (DataFileStream<GenericRecord> dataFileStream = new DataFileStream<>(inputStreamIOSupplier.get(),
-                datumReader)) {
-            // Wrap DataFileStream in a Stream using a Spliterator for lazy processing
-            return StreamSupport.stream(
-                    Spliterators.spliteratorUnknownSize(dataFileStream, Spliterator.ORDERED | Spliterator.NONNULL),
-                    false);
+        InputStream inputStream; // NOPMD CloseResource: being closed in try resources iterator
+        DataFileStream<GenericRecord> dataFileStream; // NOPMD CloseResource: being closed in try resources iterator
+        try {
+            // Open input stream from S3
+            inputStream = inputStreamIOSupplier.get();
+
+            // Ensure the DataFileStream is initialized correctly with the open stream
+            dataFileStream = new DataFileStream<>(inputStream, datumReader);
+
+            // Wrap DataFileStream in a Stream using a custom Spliterator for lazy processing
+            return StreamSupport.stream(new AvroRecordSpliterator<>(dataFileStream), false).onClose(() -> {
+                try {
+                    inputStream.close(); // Ensure the reader is closed after streaming
+                } catch (IOException e) {
+                    LOGGER.error("Error closing BufferedReader: {}", e.getMessage(), e);
+                }
+            });
         } catch (IOException e) {
             LOGGER.error("Error in DataFileStream: {}", e.getMessage(), e);
             return Stream.empty(); // Return an empty stream if initialization fails
         }
     }
 
+    private static class AvroRecordSpliterator<T> implements Spliterator<T> {
+        private final DataFileStream<GenericRecord> dataFileStream;
+
+        public AvroRecordSpliterator(final DataFileStream<GenericRecord> dataFileStream) {
+            this.dataFileStream = dataFileStream;
+        }
+
+        @Override
+        public boolean tryAdvance(final Consumer<? super T> action) {
+            try {
+                if (dataFileStream.hasNext()) {
+                    final GenericRecord record = dataFileStream.next();
+                    action.accept((T) record);
+                    return true;
+                }
+            } catch (Exception e) { // NOPMD AvoidCatchingGenericException
+                LOGGER.error("Error while reading Avro record: {}", e.getMessage(), e);
+                return false;
+            }
+            return false;
+        }
+
+        @Override
+        public Spliterator<T> trySplit() {
+            return null; // Can't split the data stream as DataFileStream is sequential
+        }
+
+        @Override
+        public long estimateSize() {
+            return Long.MAX_VALUE; // We don't know the size upfront
```

Comment:
Should this be the max record size allowed rather then Max Long as we can't process anything larger then that anyway?

Reply:
Long.MAX_VALUE is quite a big number, which can handle many gb of data. I think this is ok. 
The max record size which we want to introduce for malformed recs will be on the transformed record (output) size.

Reply:
This isn't the *record* size, but the (expected) number of records remaining in the stream.  MAX_VALUE is probably the right answer since we don't want to calculate or split.